lib/TutorialServo/TutorialServo.cpp

```diff
@@ -0,0 +1,50 @@
+
+#include "TutorialServo.h"
+
+TutorialServo::TutorialServo(PinName servoPin, float servoRangeInDegrees, float minPulsewidthInMs,
+                             float maxPulsewidthInMs)
+    : m_servoPwmOut(servoPin),
+      m_servoRangeInDegrees(servoRangeInDegrees),
+      m_minPulsewidthInMs(minPulsewidthInMs),
+      m_maxPulsewidthInMs(maxPulsewidthInMs) {
+  m_servoPwmOut.period(0.020);
+  // setting the member variables using initializer list
+  // otherwise nothing to do
+}
+
+TutorialServo::~TutorialServo() {
+  m_servoRangeInDegrees = 0;
+  m_minPulsewidthInMs   = 0;
+  m_maxPulsewidthInMs   = 0;
+
+  // felt i should include a destructor although seems trivial
+}
+
+void TutorialServo::setAngleRangeInDegrees(float degrees) {
+  m_servoRangeInDegrees = degrees;
+  // updating the member variable not sure if there is anything else to do here
+}
+
+void TutorialServo::setPulsewidthRangeInMs(float minPulsewidthMs, float maxPulsewidthMs) {
+  m_minPulsewidthInMs = minPulsewidthMs;
+  m_maxPulsewidthInMs = maxPulsewidthMs;
+  // updating the member variables (not sure if there is anything else to do here)
+}
+
+void TutorialServo::setPositionInDegrees(float degrees) {
+  if (degrees >= m_servoRangeInDegrees) {  // checking if the desired degrees is in range
+                                           // the degrees is above the possible range
+    m_servoPwmOut.pulsewidth_us(m_maxPulsewidthInMs * 1000);
+  } else if (degrees <= 0) {
+    // degrees are less than 0
```

Comment:
In most cases, servo's are able to turn CW and CCW from a center position, i.e. maxPulsewidth corresponds to the max degree it can turn in one direction, and minPulsewidth corresponds to the max degree it can turn in the other direction. So it does not make sense that when degree is negative, you immediately let the servo turn max degree in the other direction. But it also depends on how you are interpreting 0 degrees. You could make a case to treat the max degree the servo can turn in the other direction as 0, but in that case, you should probably add a commen explainning what "degree" and "range" actually represent.

Reply:
I added a couple comments, hopefully this explains things better

Reply:
let me know if it is still unclear or if there are inconsistencies in how i use range and degrees

Reply:
Let me clarify...so from what I can tell, there are two ways of how range & degrees can be interpreted.
Route1: range is the magnitude of the full range (eg. if the servo can turn 90 in both directions, then range is 180), and degrees goes from 0 to range (eg. degree = 0, servo turn max to one direction, degree = 180, servo turn max in the other direction). I think this is the route you take in this MR. But this is not what one would normally expects, because if you look at the datasheet for the servo used in the tutorial (http://www.ee.ic.ac.uk/pcheung/teaching/DE1_EE/stores/sg90_datasheet.pdf),  it says: "Position "0" (1.5 ms pulse) is middle, "90" (~2ms pulse) is middle, is all the way to the right, "-90" (~1ms pulse) is all the way to the left.". And this is also how most servos work. So you see route 1 is not really intuitive, and if you really want to take it, you should add comments explaining what degrees and range represents.

Route 2: range is the max degree the servo can turn in one direction (eg. 90), and degrees is the degree to turn in either direction (eg. -90 to +90). This maps to how the hardware actually behaves more closely and is what I personally prefers. I'm not really against using route 1, but there should be comments for explaining degrees and range...

Pls let me know if this makes more sense

Reply:
Hmm okay this makes a lot more sense now thank you for the explanation. I think I will go with route 1 as this makes more sense to me, I hope this is okay. (I can change it if you think that is best just let me know)

In order to clearly communicate what I mean by range and degrees where do you recommend I leave these comments? (currently i just have a few inside the function definition but I’m not sure if this clutters things up)

I updated some of my comments in the function definition let me know what you think.

Reply:
Yeah go ahead. I think the comments you had in the function are not really necessary (the ones you have inside the two ifs). I would recommend put explanation above the function definition in the header file.

Reply:
moved the comments and removed some useless ones